hail/python/hailtop/saige/saige.py

```diff
@@ -0,0 +1,388 @@
+                config_str = self._dump_data_config_to_yaml(
+                    mt_path, output_dir, checkpoint_dir.name, phenotypes, variant_chunks, input_data_type
+                )
+                f.write(config_str + '\n')
+
+            for phenotype_group in phenotypes:
+                for phenotype in phenotype_group:
+                    null_glmm = self.config.step1_null_glmm(
+                        b,
+                        input_bfile=input_plink_data,
+                        input_phenotypes=input_phenotypes,
+                        phenotype=phenotype,
+                        analysis_type=SaigeAnalysisType.VARIANT,
+                        covariates=covariates,
+                        user_id_col=user_id_col,
+                        output_dir=checkpoint_dir.name,
+                    )
+                    for variant_chunk in variant_chunks:
+                        self.config.step2_spa(
+                            b,
+                            mt_path=mt_path,
+                            output_dir=output_dir,
+                            analysis_type=SaigeAnalysisType.VARIANT,
+                            null_model=null_glmm,
+                            input_data_type=input_data_type,
+                            chunk=variant_chunk,
+                            phenotype=phenotype,
+                        )
+
+            run_kwargs = run_kwargs or {}
+            run_kwargs.pop('wait')
+            b_handle = b.run(**run_kwargs, wait=False)
+
+            with hfs.open(f'{output_dir}/batch_info.yaml', 'w') as f:
+                info = yaml.safe_dump({'batch_id': b_handle.id, 'name': b.name, 'attributes': b.attributes})
+                f.write(info + '\n')
+
+            b_handle.wait()
+
+            results_tables = []
+            for phenotype_group in phenotypes:
+                for phenotype in phenotype_group:
+                    results_path = self.config.step2_spa.output_dir(output_dir, phenotype)
+                    ht = hl.import_table(results_path, impute=True)
+                    ht = ht.annotate(phenotype=phenotype.name)
+                    results_tables.append(ht)
+
+            if len(results_tables) == 1:
+                results = results_tables[0]
+            else:
+                results = results_tables[0].union(*results_tables[1:])
+
+            results.checkpoint(f'{output_dir}/saige_results.ht')
+
+        return results
+
+    def run_saige_gene(
+        self,
+        *,
+        mt: Union[str, hl.MatrixTable],
+        phenotypes: Union[List[str], Phenotypes],
+        covariates: List[str],
+        group_col: str,
+        output_dir: str,
+        b: Optional[hb.Batch] = None,
+        checkpoint_dir: Optional[str] = None,
+        variant_chunks: Optional[VariantChunks] = None,
+        run_kwargs: Optional[dict] = None,
+    ):
+        if checkpoint_dir is None:
+            checkpoint_dir = hl.TemporaryDirectory()
```

Comment:
It seems like the user should be forced to explicit provide a checkpoint directory if they want to create/restore-from checkpoints and otherwise we should just not use checkpoints.

Reply:
I think I conflated working directory with checkpoint directory here. If the user wants checkpointing, then we should do the "work" in the checkpoint directory. Else, we should just blow away the working directory when we're done. I can rename to make this clearer.

Reply:
Are you thinking about the LocalBackend case? I don't understand what you mean about doing work. Intermediate files should get cleaned up by hailtop.batch's infrastructure. And it seems the rest of your infrastructure here avoids `write_output` if you're not checkpointing. What would we need to blow away? Also, what does a working directory mean when you're in the cloud?